Simplify and improve multiarch unit-tests

jschintag opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
Before this PR:
Architecture specific unit-test results are hidden in the testcode
After this PR:
Architecture specific unit-test results are separate testcases using Tables.

<!-- (optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: -->
Fixes #

### Why we need it and why it was done in this way
This improves the readability of testcases and moves into making the unit-tests more indepented of the architecture of the host it is being run on.

The following tradeoffs were made:
- 1 instance each of arm64 and ppc64le code has been left as is, since it is only a special case for them and i'm no expert on these 2 architectures, so i left it as is.
- `pkg/defaults` has been left unchanged, since the code is already in architecture specific files and it would be more complicated to use the interfaces approach here.
The following alternatives were considered:

Links to places where the discussion took place: <!-- optional: slack, other GH issue, mailinglist, ... -->

### Special notes for your reviewer
<!-- optional -->
I have split the PR into commits per package, but i can happily squash them if needed.

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [x] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [x] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [x] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [x] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [x] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

jschintag commented:
/cc @dankenigsberg

jschintag commented:
/hold

I forgot about running `make generate` and it deepcopy-gen does not like interfaces.

jschintag commented:
/unhold

0xFelix commented on pkg/virt-api/webhooks/utils.go:
```diff
@@ -77,10 +77,3 @@ type Informers struct {
 func IsARM64(vmiSpec *v1.VirtualMachineInstanceSpec) bool {
```
Can this be dropped too?

0xFelix commented on pkg/virt-launcher/virtwrap/api/archdefaults_test.go:
```diff
@@ -0,0 +1,37 @@
+package api
+
+import (
+	ginkgo "github.com/onsi/ginkgo/v2"
```
Should be okay to dot import this

jschintag replied:
No, there is already an Entry function in the package. Other tests here import it this way as well.

0xFelix commented on pkg/virt-handler/node-labeller/arch_labeller.go:
```diff
@@ -0,0 +1,77 @@
+var _ = archLabeller(&defaultArchLabeller{})
+
+type archLabeller interface {
+	ShouldLabelNodes() bool
```
Why do you mix exported/unexported methods in this interface?

jschintag replied:
This one Method is needed outside of the package, the rest is only used internally.

0xFelix commented on pkg/virt-controller/services/template.go:
```diff
@@ -316,14 +314,6 @@ func (t *templateService) IsPPC64() bool {
 	return t.clusterConfig.GetClusterCPUArch() == "ppc64le"
```
Still required?

jschintag replied:
Yes, ppc64le has some special cases in the templates. The rest of the arches was not used.

0xFelix commented on pkg/virt-handler/node-labeller/cpu_plugin.go:
```diff
@@ -82,15 +81,15 @@ func (n *NodeLabeller) loadDomCapabilities() error {
 			}
 
 			n.cpuModelVendor = mode.Vendor.Name
 			// On s390x the xml does not include a CPU Vendor, however there is only one company selling them anyway.
```
Comment does not relate to code below anymore?

jschintag replied:
Yes, i'll move the comment into the s390x.go file instead.

jschintag commented on pkg/virt-handler/node-labeller/cpu_plugin.go:
```diff
@@ -82,15 +81,15 @@ func (n *NodeLabeller) loadDomCapabilities() error {
 	for _, mode := range hostDomCapabilities.CPU.Mode {
 		if mode.Name == v1.CPUModeHostModel {
-			if virtconfig.IsARM64(n.arch) {
+			if !n.Arch.supportsHostModel() {
```
I guess it can be more general, though in practice ARM is currently the only arch we support were this is not the case.

jschintag commented on pkg/virt-launcher/virtwrap/api/defaults.go:
```diff
@@ -8,54 +8,26 @@ const (
-
-func (d *Defaulter) isS390X() bool {
-	return d.Architecture == "s390x"
+	Architecture  string
```
You are right, i left it because it was exported, but it seems it was only used internally anyway.
Let me run a quick unit-test to check it does not break anything, then i'll push the changes.

jschintag replied:
I dropped the field.

jschintag commented on cmd/virt-handler/virt-handler.go:
```diff
@@ -312,7 +311,7 @@ func (app *virtHandlerApp) Run() {
 
 	// Node labelling is only relevant on x86_64 and s390x arches.
-	if virtconfig.IsAMD64(runtime.GOARCH) || virtconfig.IsS390X(runtime.GOARCH) {
+	if nodeLabellerController.Arch.ShouldLabelNodes() {
```
Done

EdDev reviewed (changes requested):
Thank you for investing in this.

I have two high-level comments about this change:
- It will be better to split the vir-launcher changes ("api") from the virt-handler ones (node-labeller). These are not easy changes and require a more focused context.
To be specific, I mean different PRs.
- In regards to the `virtwrap/api` package, it is unclear to me why it cannot be part of the `converter/arch`. Can the logic here be encorporated in to the converter which is already split? Both eventually represent the creation of the domain spec and having it fragmented is harder to maintain and understand.

Regarding the node-labeller, it seem to me like the conditions have been preserved which IMO could have been reduced by creating for each architecture its own logic. As it appears now, if a new arch is added or changes are needed to existing ones, we will need to touch the core/general code base (and we do not want to have that).
But we should defer this handling to a separate PR so we can focus on that.

jschintag commented:
Thank you for your review @EdDev.

I have a few questions/points:
- Regarding your 1. point, do i get it right that you want to split it into 3 PRs?
    1. node-labeller (virt-handler)
    2. virtwrap/api (virt-launcher)
    3. Other (The 3 smaller commits left)
- Regarding the 2. point, i haven't thought about doing it that way, but looking at it, that is not going to work. Unfortunately `converter/arch` imports `virtwrap/api`, so we have the problem of cyclic imports.
- Regarding node-labeller, looking at it, i think i can maybe move more parts of the logic into the arch specific files. However this is always a balancing act between duplicating code and having separate code for arches. In general, changing core/general code when adding a new architecture or if there are changes in an arch is unavoidable, because new arches especially will not always need special handling at the same place. There is a lot of overlap, but it is not 100%. Same reason there are isARM64/isPPC64LE functions left after this PR.

jschintag commented:
Hi @EdDev,
could you clarify the questions (https://github.com/kubevirt/kubevirt/pull/13852#issuecomment-2647531318) i had regarding how you want the changes?

EdDev commented:
> Thank you for your review @EdDev.
> 
> I have a few questions/points:
> 
>     * Regarding your 1. point, do i get it right that you want to split it into 3 PRs?
>       
>       1. node-labeller (virt-handler)
>       2. virtwrap/api (virt-launcher)
>       3. Other (The 3 smaller commits left)

Yes.
The "other" commits are small cleanups of unused code and adjustments.
The main separation is in the first two.


>     * Regarding the 2. point, i haven't thought about doing it that way, but looking at it, that is not going to work. Unfortunately `converter/arch` imports `virtwrap/api`, so we have the problem of cyclic imports.

Can the "defaulter" be extracted to its own package?

>     * Regarding node-labeller, looking at it, i think i can maybe move more parts of the logic into the arch specific files. However this is always a balancing act between duplicating code and having separate code for arches. In general, changing core/general code when adding a new architecture or if there are changes in an arch is unavoidable, because new arches especially will not always need special handling at the same place. There is a lot of overlap, but it is not 100%. Same reason there are isARM64/isPPC64LE functions left after this PR.

Yes, it may be unavoidable. However, it can be directed into that direction.
Some implication can be solved by helper functions called from the arch specific implementation.
IMO it will be easier to discuss this and move this valuable contribution forward if we'll have the smaller PRs.
But I agree it can only come if we'll find an agreed path to this.

Sorry for the late response, I'm too slow these days. I will try to be more responsive.
If you need help with something specific, please feel free to contact me offline as well.

jschintag commented:
As requested i have split this PR in 3 parts:
- node-labeller: #13973
- virtwrap/api: #13974
- Rest (this PR)

Additionally i have, as requested, moved the arch defaulter from api into a separate package. Also i simplified the code of node-labeller a tiny bit by moving more logic into the arch labeller.

jschintag commented:
Hmm, i forgot about this. But we need to merge https://github.com/kubevirt/kubevirt/pull/13973 before we merge this PR, otherwise it just gives merge conflicts in the future and extra work.

EdDev commented on pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go:
```diff
@@ -204,13 +204,14 @@ var _ = Describe("VirtualMachineInstance Mutator", func() {
-		if webhooks.IsPPC64(vmiSpec) {
-			Expect(vmiSpec.Domain.Machine.Type).To(Equal("pseries"))
-		} else if webhooks.IsARM64(vmiSpec) {
+		switch vmiSpec.Architecture {
```
This is not directly related to the use of `switch`, but it is odd to add conditions like this here.
The input of `arch` and an expected "Type" should be passed in so the assertion could be done without these conditions. Similar to what was done in the vm-mutator tests.

jschintag replied:
I needed to look through it again, but the reason is because the "Type" is not known for the last test:
```
Entry("when architecture is not specified", "", v1.DefaultCPUModel))
```

0xFelix replied:
+1

I suggest to remove the switch and to add another arg to the table's func that contains the expected machine type. Similar tt how it is done with the cpu model already.

jschintag replied:
As i said [here](https://github.com/kubevirt/kubevirt/pull/13852/files#r1959291298), the type is not known for the last testcase. This depends then on the architecture of the host.
To make the type another parameter, i would need to drop the unknown architecture testcase.

jschintag replied:
I have had some time to follow the code that sets the machineType and found a solution.

> Not the prettiest, but the test case below overrides `webhooks.Arch`. Would that be an option? This way you can predict the default architecture.

Unfortunately that value is not used.

What i found that works is, to use the same function that is called as part of defaults to set the machineType.
```
mutator.ClusterConfig.GetMachineType(rt.GOARCH)
```
This at least removes the switch from the test and essentially uses the switch in `virt-config` instead. This way we test it at least calls the function and we make sure it works correctly by testing it with the specific arches first.
From my side, i think this does not change what the test covers so it should be good.

EdDev reviewed (commented):
There seems to be a problem with the build.
Please run `make test` and `make lint` and see it is passing.

Here is what I can see at the moment:
```
 13:14:47: pkg/virt-handler/node-labeller/cpu_plugin.go:1: : # kubevirt.io/kubevirt/pkg/virt-handler/node-labeller [kubevirt.io/kubevirt/pkg/virt-handler/node-labeller.test]
13:14:47: pkg/virt-handler/node-labeller/cpu_plugin.go:92:18: undefined: virtconfig.IsS390X
13:14:47: pkg/virt-handler/node-labeller/cpu_plugin.go:141:52: undefined: virtconfig.IsS390X
13:14:47: pkg/virt-handler/node-labeller/node_labeller.go:160:16: undefined: virtconfig.IsAMD64
13:14:47: pkg/virt-handler/node-labeller/node_labeller.go:160:46: undefined: virtconfig.IsS390X (typecheck) 
```

EdDev commented:
> Hmm, i forgot about this. But we need to merge #13973 before we merge this PR, otherwise it just gives merge conflicts in the future and extra work.

Ohh so this is why it fails the build?

jschintag commented:
> > Hmm, i forgot about this. But we need to merge #13973 before we merge this PR, otherwise it just gives merge conflicts in the future and extra work.
> 
> Ohh so this is why it fails the build?

Yes, i should probably add
/hold
to make that more clear.

jschintag commented:
I have rebased on top of the merged #13973 . CI should now be able to build it.

/unhold

jschintag commented:
I have rebased the PR on current main.

0xFelix commented on pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter_test.go:
```diff
@@ -483,13 +483,15 @@ var _ = Describe("Validating VMICreate Admitter", func() {
 			vmi := api.NewMinimalVMI("testvmi")
-			if webhooks.IsPPC64(&vmi.Spec) {
+
+			switch vmi.Spec.Architecture {
```
This test depends on the arch of the host, right? Can you make it arch independent? I.e. converting it to table?

jschintag replied:
Will do

jschintag replied:
Having looked at the code, it seems the vmi Architecture field was never actually set, so the test just checked amd64 anyway, making the whole if/switch unnecessary.
I have converted it to a Table and fixed the test such that different arches are actually tested.

jschintag replied:
I have also found out that the validation of the machine type is not quite working as expected, since everything which contains `q35` seems to be valid, including `reiterithq35`. I will check that out further and open an issue if required.

0xFelix commented on pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go:
```diff
@@ -192,37 +192,26 @@ var _ = Describe("VirtualMachineInstance Mutator", func() {
+		Entry("when architecture is arm64", "arm64", v1.CPUModeHostPassthrough, "virt"),
+		Entry("when architecture is ppc64le", "ppc64le", v1.DefaultCPUModel, "pseries"),
+		Entry("when architecture is s390x", "s390x", v1.DefaultCPUModel, "s390-ccw-virtio"),
+		Entry("when architecture is not specified", "", v1.DefaultCPUModel, ""),
```
Won't that break on arm since it is using `v1.DefaultCPUModel` instead of `v1.CPUModeHostPassthrough` now? Ideally our tests would not depend on the host architecture at all. I don't see which value this last test case has?

jschintag replied:
Right, i forgot about arm...

As for the value, it checks that we can deal with an unknown Architecture without error. I think that is the best way to test it, since most VMs will likely not have the arch or machine type set and rely on this code path.

jschintag replied:
Wait we are not running any arm64 unit tests in the CI? I don't see a test lane below for that?

0xFelix replied:
Can you extract the last test case into another table where you set the default architecture in the cluster configuration, so you can predict what should happen? This way we can get rid of the `rt.GOARCH` dependency and test all possible arch combinations independent of the architecture of the host executing the tests (which IMO is pretty bad). It could be in a follow up.

jschintag replied:
Yes, i can do that as a follow up.

jschintag commented:
@0xFelix Could you please approve the PR so it gets merged? Than i can get to implementing the follow up.

0xFelix commented:
@jschintag Not sure I'm able to. I think we need a core approver for that.

/approvve

0xFelix commented:
@jschintag Not sure I'm able to. I think we need a core approver for that.

/approve

jean-edouard commented:
The code for this PR looks good to me, but it doesn't seem to do what the PR title/description says...
I assume it just needs to be adjusted to reflect the changes that occurred after the PR was opened.

jschintag commented:
> The code for this PR looks good to me, but it doesn't seem to do what the PR title/description says... I assume it just needs to be adjusted to reflect the changes that occurred after the PR was opened.

Thank you @jean-edouard . Yes, the major things that moved the code to separate files was moved into separate PRs. I have updated the title and description.

jean-edouard reviewed (commented):
/approve
